Typevar matching fix + implicit conversions from Scalar to int/float

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26453 Typevar matching fix + implicit conversions from Scalar to int/float**

Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was the original intent, but turned out to break existing user code
that relied on the de facto dynamic downcast. (The error only occurs when
a specialized list representation is involved.)

The current fix includes the prevention of typevar widening, but
adds implicit conversions from Scalar to float or int as needed
to satisfy a matched schema.

TODO: remove temp code that preserves old behavior for special case

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

Commit:
Typevar matching fix + implicit conversions from Scalar to int/float

Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

zdevito commented on aten/src/ATen/core/type.cpp:
```diff
@@ -351,7 +351,17 @@ MatchTypeReturn matchTypeVariables(
       return MatchTypeReturn::Success();
     } else if (auto unified = unifyTypes(it->second, actual)) {
-      type_env[vt->name()] = *unified;
+      // TODO: remove once tests that rely on old behavior have been fixed
```
The loophole can also be patched by adding a specific overload for the append case. I think this is safer because it restricts the buggy back-compat behavior to just the append case seen in the wild and not to all optional str uses.

`append(List[str] s, str? elem)`

zdevito reviewed (commented):
Looks good. We should land it as a stack with the overload changes so that we do not temporarily produce a state where scalar + scalar does integer addition.

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

bhosmer commented:
@zdevito should be good to go on tighter patch plus overload changes. There's some slight subtlety around the latter - did things as straightforwardly as seemed reasonable but LMK if you'd rather see it done differently, missing overloads etc.

zdevito reviewed (approved):
Looks good. Can we add one test for scalar addition of something to check the right overload is hit?

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

bhosmer commented:
> Looks good. Can we add one test for scalar addition of something to check the right overload is hit?

@zdevito for sure, in fact added a bit more testing (though < 100% coverage) since the overloads are straight per-op additions to the registry. The original dream was to layer the behavior on generically as like Scalar-specific smarts during schema matching that didn't complicate things too much, but things are too nonuniform to avoid op-specific awareness (e.g. lots of binops don't obey the `(lhs is int && rhs is int) ? int : float` result type rule), so it seemed nicer to keep the matching behavior agnostic.

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was per the original type system design, but turned out to break
existing user code that relied on the de facto dynamic downcast. (The
error required a specialized list representation.)

The current fix includes the prevention of typevar widening, but
adds logic to insert implicit conversions from Scalar to float or int
as needed to satisfy a matched schema.

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

bhosmer commented:
@pytorchbot retest this please

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was the original intent, but turned out to break existing user code
that relied on the de facto dynamic downcast. (The error only occurs when
a specialized list representation is involved.)

The current fix includes the prevention of typevar widening, but
adds implicit conversions from Scalar to float or int as needed
to satisfy a matched schema.

TODO: remove temp code that preserves old behavior for special case

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

[ghstack-poisoned]

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was the original intent, but turned out to break existing user code
that relied on the de facto dynamic downcast. (The error only occurs when
a specialized list representation is involved.)

The current fix includes the prevention of typevar widening, but
adds implicit conversions from Scalar to float or int as needed
to satisfy a matched schema.

TODO: remove temp code that preserves old behavior for special case

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

[ghstack-poisoned]

Commit:
Update on "Typevar matching fix + implicit conversions from Scalar to int/float"


Previously, schema matching would incorrectly widen typevar bindings
when later occurrences were supertypes of earlier ones. This allowed
callsites like `floatlist.append(tensor.item())` to pass the typechecker,
causing a runtime assert (issue #24856).

An earlier, reverted fix (#25136) insisted on strict equality across all
occurrences of a typevar, necessitating explicit casts around Scalar-typed
arguments to int- or float-typed parameters, like `tensor.item()` above.
This was the original intent, but turned out to break existing user code
that relied on the de facto dynamic downcast. (The error only occurs when
a specialized list representation is involved.)

The current fix includes the prevention of typevar widening, but
adds implicit conversions from Scalar to float or int as needed
to satisfy a matched schema.

TODO: remove temp code that preserves old behavior for special case

Differential Revision: [D17470598](https://our.internmc.facebook.com/intern/diff/D17470598)

[ghstack-poisoned]
